src/transformers/models/superpoint/modeling_superpoint.py

```diff
@@ -0,0 +1,469 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""PyTorch SuperPoint model."""
+
+from typing import Optional, Tuple, Union
+
+import torch
+from torch import nn
+
+from transformers import PreTrainedModel
+from transformers.modeling_outputs import (
+    BaseModelOutputWithNoAttention,
+    ImagePointDescriptionOutput,
+)
+from transformers.models.superpoint.configuration_superpoint import SuperPointConfig
+
+from ...pytorch_utils import is_torch_greater_or_equal_than_1_13
+from ...utils import (
+    add_code_sample_docstrings,
+    add_start_docstrings,
+    add_start_docstrings_to_model_forward,
+    logging,
+)
+
+
+logger = logging.get_logger(__name__)
+
+_CONFIG_FOR_DOC = "SuperPointConfig"
+
+_CHECKPOINT_FOR_DOC = "stevenbucaille/superpoint"
+
+
+SUPERPOINT_PRETRAINED_MODEL_ARCHIVE_LIST = ["stevenbucaille/superpoint"]
+
+
+def remove_keypoints_from_borders(
+    keypoints: torch.Tensor, scores: torch.Tensor, border: int, height: int, width: int
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Removes keypoints (and their associated scores) that are too close to the border"""
+    mask_h = (keypoints[:, 0] >= border) & (keypoints[:, 0] < (height - border))
+    mask_w = (keypoints[:, 1] >= border) & (keypoints[:, 1] < (width - border))
+    mask = mask_h & mask_w
+    return keypoints[mask], scores[mask]
+
+
+def top_k_keypoints(keypoints: torch.Tensor, scores: torch.Tensor, k: int) -> Tuple[torch.Tensor, torch.Tensor]:
+    """Keeps the k keypoints with highest score"""
+    if k >= len(keypoints):
+        return keypoints, scores
+    scores, indices = torch.topk(scores, k, dim=0)
+    return keypoints[indices], scores
+
+
+def simple_nms(scores: torch.Tensor, nms_radius: int) -> torch.Tensor:
+    """Applies non-maximum suppression on scores"""
+    if nms_radius < 0:
+        raise ValueError("Expected positive values for nms_radius")
+
+    def max_pool(x):
+        return nn.functional.max_pool2d(x, kernel_size=nms_radius * 2 + 1, stride=1, padding=nms_radius)
+
+    zeros = torch.zeros_like(scores)
+    max_mask = scores == max_pool(scores)
+    for _ in range(2):
+        supp_mask = max_pool(max_mask.float()) > 0
+        supp_scores = torch.where(supp_mask, zeros, scores)
+        new_max_mask = supp_scores == max_pool(supp_scores)
+        max_mask = max_mask | (new_max_mask & (~supp_mask))
+    return torch.where(max_mask, scores, zeros)
+
+
+class SuperPointConvBlock(nn.Module):
+    def __init__(self, in_channels: int, out_channels: int, add_pooling: bool = False) -> None:
```

Comment:
All custom layers should accept the config as their first argument

Reply:
I am not sure what to do here, because if I add config as argument, it will not be used by the module because it relies only on `in_channels`, `out_channels` and `add_pooling`. But it can't rely exclusively on config since the SuperPointConvBlock can't know what values to pick from `encoder_hidden_sizes` ? 🤔

Reply:
That's OK, it should still have it as its first argument though

Reply:
Fixed in [this commit](https://github.com/huggingface/transformers/pull/28966/commits/41ee72af5e79860da85db4089c1f0174ac3c7a42)